lib/Semantics/check-data.cpp

```diff
@@ -10,6 +10,59 @@
 
 namespace Fortran::semantics {
 
+template<typename T> void DataChecker::CheckIfConstantSubscript(const T &x) {
+  if (const auto *expr{GetExpr(x)}) {
+    if (!evaluate::IsConstantExpr(*expr)) {  // C875,C881
+      context_.Say(parser::FindSourceLocation(x),
+          "Data object must have constant bounds"_err_en_US);
+    }
+  }
+}
+
+void DataChecker::CheckSubscript(const parser::SectionSubscript &subscript) {
```

Comment:
Would it be easier or cleaner to perform expression analysis on the variable and then apply these checks to the parsed expression?  That will catch more errors (e.g., rank problems) and we'll probably want to use type-checked expressions when converting the `DATA` statement into llvm.

Reply:
@klausler ,  I am not sure about this. I will try to implement this way. I believe you are asking me to call expression analysis on parser::Variable and to implement these checks when it is analysed. Please correct me if I am wrong. If it is so, DataImpliedDo objects are either Arrayrefs or Structure Components and they do not cover all variables as per parsing rule (They will not be parsed as parser::Variable). I can look into analysing dataref (which is called already for dataImpliedDo) but I dont think it helps in catching rank related issues.

Reply:
Did you try this?  You requested another review, but I don't see any changes in the direction that I suggested.

Reply:
> Did you try this? You requested another review, but I don't see any changes in the direction that I suggested.

Sorry. My Mistake. I am trying to implement. I will soon notify with the change.

Reply:
@anchu-rajendran The lowering of DATA statements to FIR would like the values in a data statement to have been evaluated.  I believe this work is in the scope of what you are doing.  Do you have an estimate for when you might push an update?

Reply:
> @anchu-rajendran The lowering of DATA statements to FIR would like the values in a data statement to have been evaluated. I believe this work is in the scope of what you are doing. Do you have an estimate for when you might push an update?

Hi Steve, I will push an update in a week.

Reply:
> @anchu-rajendran The lowering of DATA statements to FIR would like the values in a data statement to have been evaluated. I believe this work is in the scope of what you are doing. Do you have an estimate for when you might push an update?

The value evaluation of data statements is currently not under the scope of these semantic checks. Could you please share an example where data evaluation is necessary. I will look into the same and update.